src/App.tsx

```diff
@@ -1,26 +1,254 @@
+      .createTodo({ userId, title, completed })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+      })
+      .catch(error => {
+        setErrorMessage('Unable to add a todo');
+        setErrorVisible(true);
+        setTimeout(() => {
+          setErrorVisible(false);
+        }, 3000);
+        throw error;
+      });
+  }
+
+  function deleteTodo(todoId: number) {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        );
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+
+        setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+      })
+      .catch(error => {
+        setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+        setErrorMessage('Unable to delete a todo');
+        setErrorVisible(true);
+        setTimeout(() => {
+          setErrorVisible(false);
+        }, 3000);
+        throw error;
+      });
+  }
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (!title || title.trim() === '') {
+      setErrorMessage('Title should not be empty');
+      setErrorVisible(true);
+      setTimeout(() => {
+        setErrorVisible(false);
+      }, 3000);
+
+      return;
+    }
+
+    const temporaryTodo: Todo = {
+      id: 0,
+      userId: todoService.USER_ID,
+      title: title,
+      completed: false,
+    };
+
+    setTempTodo(temporaryTodo);
+
+    if (inputRef.current) {
+      inputRef.current.disabled = true;
+    }
+
+    addTodo({
+      id: 0,
+      userId: todoService.USER_ID,
+      title: title,
+      completed: false,
+    })
+      .then(() => {
+        reset();
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setTempTodo(null);
+      })
+      .catch(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setTempTodo(null);
+      })
+      .finally();
+  };
+
+  function updateTodo(updatedTodo: Todo) {
+    setLoadingTodoIds(prevIds => [...prevIds, updatedTodo.id]);
+    todoService
+      .updateTodo(updatedTodo)
+      .then(todo => {
+        setTodos(currentTodos => {
+          const newTodo = todo;
+          const newTodos = [...currentTodos];
+          const index = newTodos.findIndex(
+            todoOpt => todoOpt.id === updatedTodo.id,
+          );
+
+          if (todo.completed === false) {
+            newTodo.completed = true;
+            newTodos.splice(index, 1, newTodo);
+
+            return newTodos;
+          } else {
+            newTodo.completed = false;
+            newTodos.splice(index, 1, newTodo);
+
+            return newTodos;
+          }
+        });
```

Comment:
you can just map the state

try to simplify it(using the .map()), or feel free to ask for some help in the chat